runtime/src/accounts.rs

```diff
@@ -314,36 +346,68 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, (account, None));
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &mut HashMap<Pubkey, Arc<AtomicU64>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
-    ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+    ) -> Result<CreditOnlyLocks> {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+
+        for k in credit_debit_keys.iter() {
+            if locks.contains(k) || credit_only_locks.contains_key(k) {
```

Comment:
Does this imply that a credit_only account for one transaction can also show up as a credit_debit account in another transaction?

Reply:
Certainly. In the case of System Transfer where From is a credit-debit account and To is a credit-only account: imagine I am paying you lamports at the same time that you are trying to pay someone else.